components/cli/src/command/finalize.rs

```diff
@@ -0,0 +1,178 @@
+    pub output_config_path: PathBuf,
+    #[structopt(long, short = "p")]
+    /// List of paths to public parts of configuration of all the nodes
+    /// in the network.
+    pub public_configs: Vec<PathBuf>,
+    #[structopt(long)]
+    /// Listen address for node public API.
+    ///
+    /// Public API is used mainly for sending API requests to user services.
+    pub public_api_address: Option<SocketAddr>,
+    #[structopt(long)]
+    /// Listen address for node private API.
+    ///
+    /// Private API is used by node administrators for node monitoring and control.
+    pub private_api_address: Option<SocketAddr>,
+    #[structopt(long)]
+    /// Cross-origin resource sharing options for responses returned by public API handlers.
+    pub public_allow_origin: Option<String>,
+    #[structopt(long)]
+    /// Cross-origin resource sharing options for responses returned by private API handlers.
+    pub private_allow_origin: Option<String>,
+}
+
+struct ValidatedConfigs {
+    common: CommonConfigTemplate,
+    public_configs: Vec<NodePublicConfig>,
+}
+
+impl Finalize {
+    fn validate_configs(public_configs: Vec<SharedConfig>) -> Result<ValidatedConfigs, Error> {
+        let mut map = BTreeMap::new();
+        let mut config_iter = public_configs.into_iter();
+        let first = config_iter.next().ok_or(format_err!(
+            "Expected at least one config in PUBLIC_CONFIGS"
+        ))?;
+        let common = first.common;
+        map.insert(first.node.validator_keys.consensus_key, first.node);
+
+        for config in config_iter {
+            if common != config.common {
+                bail!("Found config with different common part.");
+            };
+            if map
+                .insert(config.node.validator_keys.consensus_key, config.node)
+                .is_some()
+            {
+                bail!("Found duplicate consensus keys in PUBLIC_CONFIGS");
+            }
+        }
+        Ok(ValidatedConfigs {
+            common,
+            public_configs: map.values().cloned().collect(),
+        })
+    }
+
+    fn create_connect_list_config(
+        public_configs: &[NodePublicConfig],
+        secret_config: &NodePrivateConfig,
+    ) -> ConnectListConfig {
+        let peers = public_configs
+            .iter()
+            .filter(|config| {
+                config.validator_keys.consensus_key != secret_config.consensus_public_key
+            })
+            .map(|config| ConnectInfo {
+                public_key: config.validator_keys.consensus_key,
+                address: config.address.clone(),
+            })
+            .collect();
+
+        ConnectListConfig { peers }
+    }
+}
+
+impl ExonumCommand for Finalize {
+    fn execute(self) -> Result<StandardResult, Error> {
+        let secret_config: NodePrivateConfig = load_config_file(&self.secret_config_path)?;
+        let secret_config_dir =
+            std::env::current_dir()?.join(self.secret_config_path.parent().unwrap());
```

Comment:
Why don't you use `?` instead of `unwrap()`. As the return type is `Result...` ?

Reply:
Because `parent` returns `Option`, not `Result`